### PR TITLE
PS1 AKAO Format Overall Improvements

### PR DESCRIPTION
**Pull request by loveemu**
This PR replaces most of the current AkaoFormat code to support more driver versions accurately.

The updated code still does not emulate all driver behavior but is expected to parse real-world music sequences correctly.

### AKAO Format

* Add support for early AKAO sequences and instruments of Parasite Eve and before
* Change to use MIDI bank for custom instruments and drum kits for code maintainability
* Change matcher to FilegroupMatcher (previously GetIdMatcher was used)
* Fix the incorrect tuning of instruments whose fine-tuning value exceeds a semitone (Chrono Cross: Dragon God) (#124)
* Fix incorrect loop points (#125)
* Fix the bug where basic single-zone instruments are not saved when exporting SF2, unlike DLS (#126)
* Fix crash when loading sequences that do not use any custom instruments and drum kits (Chrono Cross: Dream's Creation) (#168)
* Provide more detailed and accurate information on file events

### Core

* Fix memory leaks on Save as Wav
* Set default velocity value to VGMSeq
* Allow VGMInstrSet without instruments to be accepted (a painful way to support the old AKAO)
* Add an experimental option variable to convert linear pan amplitude to GM2 sin/cos scale (This option is not used in the end, as it only changes the volume slightly, and if a song use pan and volume slides at the same time, the volume adjustment output will be indeterminate.)

## Motivation and Context

I wanted to add Final Fantasy 7 AKAO support to VGMTrans to celebrate the FF7 Remake.

When I came up with that plan, I was not familiar with the PS1 AKAO format and felt that the existing implementation of VGMTrans was somewhat inaccurate at some points. As I went through the analysis of Final Fantasy 7, I faced the need to analyze every game that used the same AKAO driver and spent months analyzing different aspects of the driver. This PR is the culmination of that.

For your information, the technical information I found through my research has been fed back to the following documents.

* [サウンドデータ (AKAO Format) - sagafrontier @ うぃき](https://w.atwiki.jp/sagafrontier/pages/43.html) (Japanese)
* [FF7/PSX/PSX Sound - Final Fantasy Inside](https://wiki.ffrtt.ru/index.php?title=FF7/PSX/PSX_Sound)

## How Has This Been Tested?

I have changed some of the shared code, but nothing should change in formats other than AKAO. I picked up and tested some songs from each of these games.

* Final Fantasy 7 (1997)
* SaGa Frontier (1997) (need to open the psflib file instead of the minipsf files - it takes a long time to load because all sequences and samples are stored in a single image)
* Front Mission 2 (1997)
* Chocobo's Mysterious Dungeon (1997) (additional loading procedure required because of the compressed sequence data)
* Parasite Eve (1998)
* Another Mind (1998)
* Chocobo Dungeon 2 (1998) (unlistenable, songs composed with multiple sample collections)
* Final Fantasy 8 (1999) (unlistenable, songs composed with multiple sample collections)
* Chocobo Racing (1999)
* SaGa Frontier 2 (1999) (no collection will be produced - the old GetIdMatcher works better for this game because the PSF image contains multiple song sequences and FilegroupMatcher is unable to handle it)
* Racing Lagoon (1999)
* Legend of Mana (1999)
* Front Mission 3 (1999)
* Chrono Cross (1999)
* Vagrant Story (2000) (unlistenable, songs composed with multiple sample collections)
* Final Fantasy 9 (2000)
* Final Fantasy Origins - Final Fantasy 2 (2000)

The following are some test cases I used:

* Final Fantasy 7: Still More Fighting (Final Fantasy 7 era, using drum kit)
* Final Fantasy 7: One-Winged Angel (Final Fantasy 7 era, using the second instrument set - not supported)
* SaGa Frontier: SaGa Frontier.psflib
* Front Mission 2: Arena Battle (Front Mission 2 / Chocobo Dungeon era, using drum kit)
* Parasite Eve: Out of Phase (Parasite Eve era, using a key-split instrument and drum kit)
* Another Mind: Another Mind (Another Mind era, using key-split instruments and drum kit)
* Chrono Cross: Dream's Creation (late version, a song without key-split instruments and drum kit)
* Chrono Cross: Dragon God (late version, using single-zone instruments, custom key-split instruments, and drum kit)
* Final Fantasy 9: Final Battle (late version, the special case for opcode 0xFE 0x07)

I also loaded almost every data format supported by VGMTrans one by one as a regression test.

## Limitations

* Songs are expected to be loaded from a PSF file that is properly tagged (the album name is used to identify the game)
    * Mainly due to incomplete detection of the file version. The heuristic-based version guessing routine cannot identify minor version differences
    * I do not think it is a good idea, but I cannot come up with any other good ideas
* Imperfections in auto-matching of song collections and its MIDI/SF2 conversion
    * FilegroupMatcher has been adapted instead of GetIdMatcher because it gives mostly good results when detecting collections from a PSF file and is easy to apply to older versions of AKAO
    * It works fine if one PSF file contains a sequence, an instrument set, and a sample collection, but otherwise, it will probably fail
    * Either way, it is difficult to always automatically link collections. Sometimes you have to create a collection manually (although the WTL GUI cannot create collections manually)
    * Some songs consist of multiple sample collections, but VGMTrans cannot properly convert those instruments to DLS/SoundFont. This is because AkaoColl's conversion code assumes there is only one sample collection
* Dummy AkaoInstrSet generated by a sequence with no custom instruments and the application design flaw behind it
    * The current codebase assigns a sequence's custom instrument set to AkaoInstrSet and an instrument file to AkaoSampColl, forcing the creation of an empty dummy AkaoInstrSet for a sequence that does not have a custom instrument
    * Ideally, an instrument file should be represented as a pair of AkaoInstrSet and AkaoSampColl and should be capable of being converted to DLS/SF2 by itself. But I did not tackle it because it would require a whole code change
* Accuracy of the parser and the MIDI/SF2 conversion
    * The complex opcodes, which do not have a one-to-one correspondence with the MIDI messages, are often not translated into the MIDI messages, often resulting in an audible difference from the original sound
       * Concretely, legato, pitch bend, vibrato, and voice overlay, and so on, respectively
    * The AkaoSeq parser does not implement some less common opcodes
        * Such opcodes are usually not used in the BGM and may require a more fine-grained identification of the driver version. Details about the unimplemented opcodes are given in the handling of event mappings in AkaoSeq.cpp
        * However, I do not think this point will be much of a problem in the normal use case
    *  The conversion support of ADSR is limited to instrument files (AkaoSampColl) only
        * The ADSR properties of key-split instruments (AkaoInstr/AkaoRgn) and drum kits (AkaoDrumKit) included in sequence files will not be converted
        * The ADSR opcodes will not be converted
    * The calculation of drum note volume is not accurate. An actual driver is supposed to ignore the original volume of the channel and recalculate based on the volume of the drum note, but VGMTrans is currently unable to emulate this
    * Drum note reverb on/off is not converted (for no technical reason if I remember correctly, it is just not implemented)
* Game-specific limitations and notes
    *  Final Fantasy 9: Dynamical branching (opcode 0xFE 0x09) in Final Battle is specifically supported by hardcoding the variable value 
    * Chocobo's Mysterious Dungeon: The sequence data are compressed, but decompression is not supported at the moment
        * The decompression routine address of the actual driver is 0x8003299C. Once the function is finished, you can extract the decompressed sequence data from RAM (PSX debugger required)
    * SaGa Frontier: Open SaGa Frontier.psflib instead of minipsfs, because minipsfs only contains the song number

## Types of changes
What types of changes does your code introduce? Put an `x` in all the boxes that apply:
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
    - I have changed some of the shared code, but nothing should change in formats other than AKAO

## Checklist
- [x] My code follows the code style of this project.
    - I am not confident about this, but I tried to keep the existing code format
- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: src/main/formats/AkaoSeq.h [hunk 1]
```diff
@@ -1,56 +1,245 @@
...
+  EVENT_FE_1F_VAGRANT_STORY
+};
 
 class AkaoSeq:
     public VGMSeq {
  public:
-  AkaoSeq(RawFile *file, uint32_t offset);
-  virtual ~AkaoSeq(void);
+  AkaoSeq(RawFile *file, uint32_t offset, AkaoPs1Version version);
+  virtual ~AkaoSeq();
```
Comment (sykhro) [inline src/main/formats/AkaoSeq.h:139]: can be removed since empty

File: src/main/formats/AkaoSeq.h [hunk 2]
```diff
@@ -1,56 +1,245 @@
...
+
+  virtual void ResetVars() override;
+  virtual bool GetHeaderInfo() override;
+  virtual bool GetTrackPointers() override;
+
+  AkaoPs1Version version() const { return version_; }
```
Comment (sykhro) [inline src/main/formats/AkaoSeq.h:145]: `[[nodiscard]]` on this and other getters

File: src/main/formats/AkaoSeq.h [hunk 3]
```diff
@@ -1,56 +1,245 @@
...
+  static uint8_t GetNumPositiveBits(uint32_t ulWord) {
+    return   ((ulWord & 0x80000000)>0) + ((ulWord & 0x40000000)>0) + ((ulWord & 0x20000000)>0) + ((ulWord & 0x10000000)>0)
+      + ((ulWord & 0x8000000)>0) + ((ulWord & 0x4000000)>0) + ((ulWord & 0x2000000)>0) + ((ulWord & 0x1000000)>0)
+      + ((ulWord & 0x800000)>0) + ((ulWord & 0x400000)>0) + ((ulWord & 0x200000)>0) + ((ulWord & 0x100000)>0)
+      + ((ulWord & 0x80000)>0) + ((ulWord & 0x40000)>0) + ((ulWord & 0x20000)>0) + ((ulWord & 0x10000)>0)
+      + ((ulWord & 0x8000)>0) + ((ulWord & 0x4000)>0) + ((ulWord & 0x2000)>0) + ((ulWord & 0x1000)>0)
+      + ((ulWord & 0x800)>0) + ((ulWord & 0x400)>0) + ((ulWord & 0x200)>0) + ((ulWord & 0x100)>0)
+      + ((ulWord & 0x80)>0) + ((ulWord & 0x40)>0) + ((ulWord & 0x20)>0) + ((ulWord & 0x10)>0)
+      + ((ulWord & 0x8)>0) + ((ulWord & 0x4)>0) + ((ulWord & 0x2)>0) + ((ulWord & 0x1));
+  }
```
Comment (sykhro) [inline src/main/formats/AkaoSeq.h:183]: it's easier to use popcount directly here. on gcc/clang it's `__builtin_popcount()`, for MSVC you can do the following;
```
#ifdef _MSC_VER
  #include <intrin.h>
  #define __builtin_popcount __popcnt
#endif
```

not sure if this works on ARM but we can think about that later on

File: src/main/formats/AkaoSeq.h [hunk 4]
```diff
@@ -1,56 +1,245 @@
...
 
  public:
-  AkaoInstrSet *instrset;
+  //AkaoInstrSet *instrset;
```
Comment (sykhro) [inline src/main/formats/AkaoSeq.h:186]: should be removed?

File: src/main/formats/AkaoSeq.h [hunk 5]
```diff
@@ -1,56 +1,245 @@
...
     : public SeqTrack {
  public:
   AkaoTrack(AkaoSeq *parentFile, long offset = 0, long length = 0);
 
-  virtual void ResetVars(void);
-  virtual bool ReadEvent(void);
+  virtual void ResetVars() override;
+  virtual bool ReadEvent() override;
+
+  AkaoSeq * seq() const {
```
Comment (sykhro) [inline src/main/formats/AkaoSeq.h:221]: `[[nodiscard]]`

File: src/main/formats/PSXSPU.h [hunk 6]
```diff
@@ -118,6 +118,16 @@ inline int RoundToZero(int val) {
...
+inline uint16_t ComposePSXADSR1(uint8_t am, uint8_t ar, uint8_t dr, uint8_t sl)
+{
+  return ((am & 1) << 15) | ((ar & 0x7f) << 8) | ((dr & 0xf) << 4) | (sl & 0x0f);
+}
+
+inline uint16_t ComposePSXADSR2(uint8_t sm, uint8_t sd, uint8_t sr, uint8_t rm, uint8_t rr)
+{
+  return ((sm & 1) << 15) | ((sd & 1) << 14) | ((sr & 0x7f) << 6) | ((rm & 1) << 5) | (rr & 0x1f);
+}
+
```
Comment (sykhro) [inline src/main/formats/PSXSPU.h:130]: minor nitpick, these can be made `constexpr`
can be skipped though, I've planned a PSXPSU refactor anyway

**Review by sykhro — CHANGES_REQUESTED** [review]
Amazing work, left a couple things to revise before merging

File: src/main/SF2File.cpp [hunk 7]
```diff
@@ -408,11 +408,15 @@ SF2File::SF2File(SynthFile *synthfile)
       sampInfo = wave->sampinfo;
     assert (sampInfo != NULL);
 
+    const int16_t coarseTune = sampInfo->sFineTune / 100;
```
[the remark this answers is not in the document]
Comment (loveemu) [reply src/main/SF2File.cpp:411]: This was not needed in the end, so I will put it back.

File: src/main/formats/AkaoFormat.h [hunk 8]
```diff
@@ -51,7 +51,8 @@ class AkaoColl:
 BEGIN_FORMAT(Akao)
   USING_SCANNER(AkaoScanner)
   //USING_MATCHER_WITH_ARG(SimpleMatcher, true)
-  USING_MATCHER_WITH_ARG(GetIdMatcher, true)
+  //USING_MATCHER_WITH_ARG(GetIdMatcher, true)
```
[the remark this answers is not in the document]
Comment (loveemu) [reply src/main/formats/AkaoFormat.h:54]: I sincerely agree

File: src/main/formats/AkaoSeq.cpp [hunk 9]
```diff
@@ -6,718 +6,1613 @@ DECLARE_FORMAT(Akao);
...
+    event_map[0xfe] = EVENT_MEASURE;
+    event_map[0xff] = EVENT_UNIMPLEMENTED;
+
+    if (version() == AkaoPs1Version::VERSION_1_1)
+    {
+      event_map[0xf3] = EVENT_F3_SAGAFRO;
+      //event_map[0xf4] = EVENT_UNIMPLEMENTED;
+      //event_map[0xf5] = EVENT_UNIMPLEMENTED;
+      //event_map[0xf6] = EVENT_UNIMPLEMENTED;
+      //event_map[0xf7] = EVENT_UNIMPLEMENTED;
```
[the remark this answers is not in the document]
Comment (loveemu) [reply src/main/formats/AkaoSeq.cpp:341]: There is no particular reason. I'll uncomment them and make them like real drivers.

File: src/main/formats/AkaoScanner.cpp [hunk 10]
```diff
@@ -10,40 +10,92 @@ AkaoScanner::~AkaoScanner(void) {
...
+
+    std::vector<AkaoInstrDatLocation> instrLocations;
+
+    if (file->GetWord(instrLocation.instrAllOffset) == 0x1010
+      && file->GetWord(instrLocation.instrDatOffset) == 0x1010)
+      instrLocations.push_back(instrLocation);
+
+    //if (file->GetWord(instr2Location.instrAllOffset) == 0x38560
+    //  && file->GetWord(instr2Location.instrDatOffset) == 0x38560)
+    //  instrLocations.push_back(instr2Location);
```
[the remark this answers is not in the document]
Comment (loveemu) [reply src/main/formats/AkaoScanner.cpp]: Uncommented and added a comment

**Comment by sykhro** [comment]
LGTM

**Comment by loveemu** [comment]
@sykhro I think I have finished correcting all the points you made.

Oh, Cirrus CI failed? I have no idea what to do. 😭

**Review by sykhro — APPROVED** [review]

**Comment by sykhro** [comment]
> @sykhro I think I have finished correcting all the points you made.
> 
> Oh, Cirrus CI failed? I have no idea what to do. sob

Everything is good, Github Actions failed due to something unrelated; AppVeyor passed and that's what matters right now.
How do we want to go about merging? It's a lot of commits but with squashing it might be harder to trace back possible bugs. If it's fine for you we can go for "rebase and merge"

**Comment by loveemu** [comment]
@sykhro I will create a merge commit if it's fine. Since I forked this branch a long time ago, I'm not sure if the rebase will run without problems.

But before that, there is one problem. I tried to rebuild this branch using Visual Studio 2019 on my local machine, but MAMELoader.h told me that unzip.h was missing. Do you have any guesses as to what I'm doing wrong?

**Comment by sykhro** [comment]
> @sykhro I will create a merge commit if it's fine. Since I forked this branch a long time ago, I'm not sure if the rebase will run without problems.

That's fine for me

> But before that, there is one problem. I tried to rebuild this branch using Visual Studio 2019 on my local machine, but MAMELoader.h told me that unzip.h was missing. Do you have any guesses as to what I'm doing wrong?

That's strange since it passed on AppVeyor. I have made some changes to zlib and minizip on `master`, it _should_ solve by itself after merging

**Comment by loveemu** [comment]
> I tried to rebuild this branch using Visual Studio 2019 on my local machine, but MAMELoader.h told me that unzip.h was missing.

This happened in the master branch as well, so it seems that it was not introduced by my change.

I will therefore proceed with making a merge commit. Thanks for the quick and good feedback.

**Comment by sykhro** [comment]
> > I tried to rebuild this branch using Visual Studio 2019 on my local machine, but MAMELoader.h told me that unzip.h was missing.
> 
> This happened in the master branch as well, so it seems that it was not introduced by my change.
> 
> I will therefore proceed with making a merge commit. Thanks for the quick and good feedback.

You're welcome, not sure what's going on there as AppVeyor builds correctly. Maybe it's something cached? Let's see what happens now with the merge

**Comment by loveemu** [comment]
> WTL: Change the code analysis rule set from AllRules to NativeRecommendedRules

I forgot about this. If it is not appropriate, you may revert to the previous version. This was only changed because I was confused by too many markers when I was fixing the code in Visual Studio.

**Comment by ipatix** [comment]
Is it possible that this PR broke envelopes of instrument banks for SF2 exports? Because I've tried a couple FF IX files and they will always result in the same envelope:

Attack 0.0 s
Decay: 4.039 s
Sustain 0 dB
Release 0.029 s

I know that it's sometimes difficult to map different types of curves to others but this does not look right.

**Comment by sykhro** [comment]
> 

Completely missed this, will look into it

**Pull request merged**